wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -887,47 +889,86 @@ Instance startMachine(LineConsumer machineLogger,
+                                                                                        null,
+                                                                                        null));
                 }
             }
         }
-        return serviceWithCorrectSource;
+        return serviceWithNormalizedSource;
     }
 
-    private Machine getMachineWithCorrectSource(MachineImpl machine, MachineSource machineSource) {
-        Machine machineWithCorrectSource = machine;
+    private Machine getMachineWithNormalizedSource(MachineImpl machine, MachineSource machineSource) {
+        Machine machineWithNormalizedSource = machine;
         if (machineSource != null) {
-            machineWithCorrectSource = MachineImpl.builder()
-                                                  .fromMachine(machine)
-                                                  .setConfig(MachineConfigImpl.builder()
-                                                                              .fromConfig(machine.getConfig())
-                                                                              .setSource(machineSource)
-                                                                              .build())
-                                                  .build();
+            machineWithNormalizedSource = MachineImpl.builder()
+                                                     .fromMachine(machine)
+                                                     .setConfig(MachineConfigImpl.builder()
+                                                                                 .fromConfig(machine.getConfig())
+                                                                                 .setSource(machineSource)
+                                                                                 .build())
+                                                     .build();
+
+        }
+        return machineWithNormalizedSource;
+    }
+
+    /**
+     * Replaces linked to this service's name with container name which represents the service in links section.
+     * The problem is that a user writes names of other services in links section in compose file.
+     * But actually links are constraints and they values should be names of containers (not services) to be linked.
+     * <br/>
+     * For example: serviceDB:serviceDbAlias -> container_1234:serviceDbAlias <br/>
+     * If alias is omitted then service name will be used.
+     *
+     * @param serviceToStart
+     *         service which is starting now
+     * @param services
+     *         all services in environment
+     */
+    @VisibleForTesting
+    void replaceServiceToContainerNameInLinks(CheServiceImpl serviceToStart, Map<String, CheServiceImpl> services) {
+        final List<String> containerLinks = serviceToStart.getLinks();
+        for (int i = 0;  i < containerLinks.size(); i++) {
+            String link = containerLinks.get(i);
```

Comment:
Taking into account, that we have only one loop in the method and this loop actually is method's body and body of this loop cannot be reused somewhere and if so, we should pass iterator and links list into it as parameters, IMO, there is no point to extract loop body into another method. @garagatyi WDYT?